File: frontend/src/types/reservation.ts
// Phase 6 — Reservation types and interfaces

export enum ReservationStatus {
  pending = "pending",
  confirmed = "confirmed",
  completed = "completed",
  cancelled = "cancelled",
}

export interface DealMinimal {
  id: string;
  product_name: string;
  deal_price: number;
  original_price: number;
  discount_pct: number;
  expiry_date: string;
  product_image_url: string | null;
}

export interface StoreMinimalRes {
  id: string;
  name: string;
  address: string;
  lat: number;
  lng: number;
}

export interface Reservation {
  id: string;
  deal_id: string;
  consumer_id: string;
  store_id: string;
  quantity: number;
  status: ReservationStatus;
  hold_expires_at: string; // ISO datetime
  reserved_at: string; // ISO datetime
  confirmed_at: string | null;
  completed_at: string | null;
  deal: DealMinimal;
  store: StoreMinimalRes;
}

export interface ReservationCreateRequest {
  deal_id: string;
  quantity: number;
}

export interface ReservationListResponse {
  items: Reservation[];
  total: number;
}

export interface ReservationStatusUpdate {
  status: "confirmed" | "completed" | "cancelled";
}

File: backend/app/schemas/reservation.py
"""Reservation Pydantic v2 schemas."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, field_validator


class ReservationCreateRequest(BaseModel):
    deal_id: str
    quantity: int = 1

    @field_validator("quantity")
    @classmethod
    def qty_ge_1(cls, v: int) -> int:
        if v < 1:
            raise ValueError("quantity must be >= 1")
        return v


class DealMinimal(BaseModel):
    id: str
    product_name: str
    deal_price: float
    original_price: float
    discount_pct: int
    expiry_date: str
    product_image_url: str | None = None

    model_config = ConfigDict(from_attributes=True)


class StoreMinimal(BaseModel):
    id: str
    name: str
    address: str
    lat: float
    lng: float

    model_config = ConfigDict(from_attributes=True)


class ReservationOut(BaseModel):
    id: str
    deal_id: str
    consumer_id: str
    store_id: str
    quantity: int
    status: str
    hold_expires_at: str
    reserved_at: str
    confirmed_at: str | None = None
    completed_at: str | None = None
    deal: DealMinimal
    store: StoreMinimal

    model_config = ConfigDict(from_attributes=True)


class ReservationListResponse(BaseModel):
    items: list[ReservationOut]
    total: int


class ReservationStatusUpdate(BaseModel):
    status: str

    @field_validator("status")
    @classmethod
    def valid_status(cls, v: str) -> str:
        allowed = {"confirmed", "completed", "cancelled"}
        if v not in allowed:
            raise ValueError(
                f"status must be one of: {', '.join(sorted(allowed))}"
            )
        return v
